### PR TITLE
Fix for #451

### PR DESCRIPTION
**Pull request by benjaminlaib**
@angrykoala 

What do you think about this approach? I tried swallowing as little information as possible while still keeping the correct error messages. I think the `waitFor*` functions using `waitFor` should all pass the error message on to the new error, not doing so is dangerous IMHO.

Also I am checking for the error being a `TimeoutError` by checking it against `puppeteer.errors.TimeoutError`.

**Edit:** I have not written tests yet, I will do so later today.

### CODE REVIEW

**Comment by angrykoala** [comment]
Thanks for your PR. Your approach seems to be in the good direction (and nice one on removing the DOMException check, it was painfully ugly) I'll read it later when I have a bit of time.

About the tests, it should be enough with copy-pasting [this test](https://github.com/angrykoala/wendigo/blob/master/tests/browser/wait_for.test.js#L168) but using the other methods to ensure errors are not being swallowed

**Comment by benjaminlaib** [comment]
Thanks for replying this quickly! Yes, I think my code is not perfect yet, probably in some edge cases there is still some information that gets swallowed by the `catch`.

What do you think about modifying the `Error` child classes so they can accept the original error in their constructor? This way the constructor could attach the error message to `this.message`, and the extra message to `this.extraMessage`, and no information would get swallowed.

We could also make the constructor check `err` to update its `this.extraMessage` if `message` is not passed. Looking at puppeteer's source there are a lot of cases where simple `Error` objects are thrown with a custom error message - in fact, `TimeoutError` is the only one that is not a simple `Error` object -, which Wendigo could check against. This would result in less logic in the `catch`-es.

You are correct in that checking on the error message is painfully ugly, however since puppeteer almost always hrows simple `Errors`, there is no way around it :(

**Comment by benjaminlaib** [comment]
I just realized a huge mistake I made: I did not check for `typeof selector === 'function'` [here](https://github.com/angrykoala/wendigo/pull/453/files#diff-790015b20bdd199743dead1150722ddeR36), so now the function body is stuffed into the error message. I will fix this later.

I noticed a design flaw with the way errors are currently handled by Wendigo after further evaluating how to properly fix the issue:

The original error message is always swallowed. I get that this is in order to provide additional information about the function whose call resulted in the Error being thrown, however the way this is currently implemented is impractical IMHO. The error message should never be swallowed, so issues like #451 can't arise.

Currently the implementation of `WendigoError` applies the same message to `err.message` (with a `[${fn}] ` prefix) and `err.extraMessage` (without prefix). Instead of this, what would you think about setting the `message` property to our custom error message, and the `extraMessage` property to the original error's message? We could just pass the original error to the `WendigoError` constructor and let the constructor do the logic.

**Comment by angrykoala** [comment]
Okay, there is a lot to unpack here:

First 


```
What do you think about modifying the Error child classes so they can accept the original error in their constructor? This way the constructor could attach the error message to this.message, and the extra message to this.extraMessage, and no information would get swallowed.
```
Sounds good, just keep in mind that some errors will be new, so the err parameter should be optional (I'm thinking maybe: new WendigoError(err, extra message)

**Comment by angrykoala** [comment]
Okay, there is a lot to unpack here:

First, your proposed approach to handling errors sound good 👍  (tbh, I haven't put an extremely good effort on error handling, so spending some time improving that sounds like a great idea). About using a constructor with error, as long as you keep in mind that this error parameter should be optional (to be able to create new errors without a previous error) sounds good to me.

About not shallowing errors: Take a look to OverrideErrors decorator, which is intended to avoid exactly that, maybe it can be useful for this rework of errors.

I have tried to cover as many error messages as possible on the unit tests (given that this is a testing library, error messages are important) so you can feel free to make modification to how stuff works and be sure that everything works as it should.

I'll review your PR as it is right now (although, I understand that you do not intent to merge it right now). I would appreciate if you want to make these changes to error handling 😄. If possible, try to split the work in several PRs, as this may be quite big to tackle on one go (but do as you prefer)

I hope I didn't miss anything, if so please let me know and ask any questions you have on the reopened issue
thanks

File: lib/browser/mixins/browser_wait.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import { errors as puppeteerErrors } from 'puppeteer';
```
Comment (angrykoala) [inline lib/browser/mixins/browser_wait.ts]: Try to wrap these errors inside [puppeteer_wrapper](https://github.com/angrykoala/wendigo/tree/master/lib/puppeteer_wrapper) feel free to add them to puppeteer_types or create a new file, the idea is that anything that is from puppeteer goes through an interface to simplify core changes when Puppeteer is updated or, eventually, use a different engine if needed

File: lib/browser/mixins/browser_wait.ts [hunk 2]
```diff
@@ -26,13 +27,15 @@ export default abstract class BrowserWait extends BrowserNavigation {
...
-            } else {
+            if (err instanceof puppeteerErrors.TimeoutError) {
                 let errMsg;
                 if (typeof selector === 'function') errMsg = `Waiting for function to return true`;
                 else errMsg = `Waiting for element "${selector}"`;
                 throw new TimeoutError("waitFor", errMsg, timeout);
+            } else if (err instanceof Error && err.message.match(/DOMException\:/)) { // TODO: move to a helper/wrapper
+                throw new QueryError("waitFor", `Invalid selector "${selector}".`);
+            } else {
+                throw new QueryError("waitFor", "An unknown error occurred: " + err.message)
```
Comment (angrykoala) [inline lib/browser/mixins/browser_wait.ts]: Maybe this shouldn't be a QueryError (WendigoError or FatalError)

Also, maybe don't add the message `An unknown error occurred: `, it doesn't really add any info

File: lib/browser/mixins/browser_wait.ts [hunk 3]
```diff
@@ -46,7 +49,10 @@ export default abstract class BrowserWait extends BrowserNavigation {
                 return !WendigoUtils.isVisible(element);
             }, timeout, selector);
         } catch (err) {
-            throw new TimeoutError("waitUntilNotVisible", `Waiting for element "${selector}" to not be visible`, timeout);
+            if (err instanceof TimeoutError) {
+                throw new TimeoutError("waitUntilNotVisible", `Waiting for element "${selector}" to not be visible`, timeout);
```
Comment (angrykoala) [inline lib/browser/mixins/browser_wait.ts]: Consider using @overrideError decorator (I'm not sure why I didn't added it here, maybe there was some reason I don't see now, if the decorator doesn't work, this approach seems good enough

File: lib/browser/mixins/browser_wait.ts [hunk 4]
```diff
@@ -154,7 +172,10 @@ export default abstract class BrowserWait extends BrowserNavigation {
                 return value === null;
             }, timeout, selector);
         } catch (err) {
-            throw new TimeoutError("waitUntilEnabled", `Waiting for element "${selector}" to be enabled`, timeout);
+            if (err instanceof TimeoutError) {
```
Comment (angrykoala) [inline lib/browser/mixins/browser_wait.ts]: Same on all these checks, Ideally, the decorator should handle this, if possible

**Comment by angrykoala** [comment]
@benjaminlaib any updates on this?

**Pull request closed without merging**